### PR TITLE
core: fix cpu_spin_trylock() against debug lock counting

### PR DESCRIPTION
**Pull request by etienne-lms**
i found this reading Joakim updates on docs.
I think we missed this...  seems we don't have any test that could stress `trylock`.

### CODE REVIEW

**Comment by etienne-lms** [comment]
`Tested-by: Etienne Carriere <etienne.carriere@linaro.org> (qemu, b2260)`

**Comment by igoropaniuk** [comment]
Maybe a bit unrelated to this PR question , but what was the primary purpose of spinlock_count_incr/spinlock_count_dec. It's was an attempt to add some light implementation of spinlocks statistics (something like lockdep in the linux kernel)?

**Comment by etienne-lms** [comment]
@igoropaniuk, spinlock counting is not a lock statistic feature. it is rather a debug helper to insure no spinlock is locked when non-secure interrupts are enabled. check use of `assert_have_no_spinlock()` in the core.

File: core/arch/arm/include/kernel/spinlock.h [hunk 1]
```diff
@@ -69,7 +69,7 @@ static inline unsigned int cpu_spin_trylock(unsigned int *lock)
 
 	assert(thread_irq_disabled());
 	locked = __cpu_spin_trylock(lock);
-	if (locked)
+	if (!locked)
```
Comment (jenswi-linaro) [inline core/arch/arm/include/kernel/spinlock.h]: This is confusing. Perhaps `cpu_spin_trylock()` should return a `bool` instead to align better with `mutex_trylock()`.
Then I'm also wondering what is really returned by the AArch64 version of `__cpu_spin_trylock()`.

Comment (jenswi-linaro) [inline core/arch/arm/include/kernel/spinlock.h]: Yes, please.
Well, take a look in `spin_lock_a64.S` : `__cpu_spin_trylock()`. What's in `x0` when it returns?

Comment (etienne-lms) [inline core/arch/arm/include/kernel/spinlock.h]: oohh! my mistake, you're right :) `x0` remains to the lock address.
I'll propose a new change, overriding my current proposal.

Comment (jenswi-linaro) [inline core/arch/arm/include/kernel/spinlock.h]: Thanks.

**Comment by jforissier** [comment]
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/arch/arm/kernel/spin_lock_a64.S [hunk 2]
```diff
@@ -74,10 +74,13 @@ END_FUNC __cpu_spin_lock
...
 .loop:	ldaxr	w1, [x0]
-	cbnz	w1, .cpu_spin_trylock_out
+	cbnz	w1, .failed
 	stxr	w1, w2, [x0]
 	cbnz	w1, .loop
-.cpu_spin_trylock_out:
+	mov	x0, #0
+	ret
+.failed:
+	mov	x0, #1
```
Comment (etienne-lms) [inline core/arch/arm/kernel/spin_lock_a64.S]: i think i'll swap the sign-off and reviewed-by tags :)>  
i update with your proposal.

Comment (jenswi-linaro) [inline core/arch/arm/kernel/spin_lock_a64.S]: Sure, fine with me. :-)

Comment (jforissier) [inline core/arch/arm/kernel/spin_lock_a64.S]: LGTM, my R-b still holds.

**Comment by jforissier** [comment]
@etienne-lms you should probably keep your S-o-b: in the first patch (even though you added a Reviewed-by) because "_The Signed-off-by: tag indicates that the signer was involved in the development of the patch, or that he/she was in the patch's delivery path._"

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
@jenswi-linaro. ok regarding s-o-b. I felt i had to add yours, but since that prevents you from reviewing tag, i removed my s-o-b to add my r-b.

i update tags and push back.

**Comment by etienne-lms** [comment]
fix tags + applied tags + fix prototype inconsistency.

**Comment by jforissier** [comment]
Rebased and merged.

**Pull request closed without merging**